File: driver/transport/wire.go
package transport

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

// ErrUnsupportedVersion is returned when a key-state message's version byte
// does not match ProtocolVersion, instead of guessing at field offsets.
var ErrUnsupportedVersion = errors.New("transport: unsupported protocol version")

// MessageType identifies which device→host message a frame carries. See
// the type registry in the "Framing" section of docs/wire-protocol.md.
type MessageType byte

const (
	MessageTypeEvent      MessageType = 1
	MessageTypeAudioChunk MessageType = 2
	MessageTypePong       MessageType = 3
	MessageTypeTrace      MessageType = 4
)

const (
	keyStateSize    = 6
	eventSize       = 10
	pongSize        = 1
	traceRecordSize = 12
	frameHeaderSize = 3 // type + payload length, little-endian uint16
)

func encodeKeyState(w io.Writer, ks KeyState) error {
	if ks.Version != ProtocolVersion {
		return fmt.Errorf("%w: got %d, want %d", ErrUnsupportedVersion, ks.Version, ProtocolVersion)
	}
	buf := make([]byte, keyStateSize)
	buf[0] = ks.KeyIndex
	buf[1] = ks.Version
	binary.LittleEndian.PutUint16(buf[2:4], ks.Color)
	buf[4] = ks.EmojiID
	if ks.Blink {
		buf[5] = 1
	}
	_, err := w.Write(buf)
	return err
}

func decodeKeyState(r io.Reader) (KeyState, error) {
	buf := make([]byte, keyStateSize)
	if _, err := io.ReadFull(r, buf); err != nil {
		return KeyState{}, err
	}
	ks := KeyState{
		KeyIndex: buf[0],
		Version:  buf[1],
		Color:    binary.LittleEndian.Uint16(buf[2:4]),
		EmojiID:  buf[4],
		Blink:    buf[5] != 0,
	}
	if ks.Version != ProtocolVersion {
		return KeyState{}, fmt.Errorf("%w: got %d, want %d", ErrUnsupportedVersion, ks.Version, ProtocolVersion)
	}
	return ks, nil
}

// writeFrame writes one device→host frame: a type byte, a little-endian
// uint16 payload length, then the payload itself. See "Framing" in
// docs/wire-protocol.md.
func writeFrame(w io.Writer, t MessageType, payload []byte) error {
	header := make([]byte, frameHeaderSize)
	header[0] = byte(t)
	binary.LittleEndian.PutUint16(header[1:3], uint16(len(payload)))
	if _, err := w.Write(header); err != nil {
		return err
	}
	if len(payload) == 0 {
		return nil
	}
	_, err := w.Write(payload)
	return err
}

// readFrame reads one frame's header and exactly the payload bytes its
// length declares, whether or not t is a known MessageType. A short read
// on either the header or the payload returns io.ErrUnexpectedEOF (or
// io.EOF at a clean boundary before the header), never a partial value.
func readFrame(r io.Reader) (t MessageType, payload []byte, err error) {
	header := make([]byte, frameHeaderSize)
	if _, err := io.ReadFull(r, header); err != nil {
		return 0, nil, err
	}
	n := binary.LittleEndian.Uint16(header[1:3])
	payload = make([]byte, n)
	if n > 0 {
		if _, err := io.ReadFull(r, payload); err != nil {
			return 0, nil, err
		}
	}
	return MessageType(header[0]), payload, nil
}

// decodeMessage decodes a frame's payload into a Message. ok is false when
// t is not a known MessageType; the caller has already consumed the
// payload by its declared length and should read the next frame.
func decodeMessage(t MessageType, payload []byte) (msg Message, ok bool, err error) {
	switch t {
	case MessageTypeEvent:
		ev, err := decodeEvent(payload)
		if err != nil {
			return Message{}, false, err
		}
		return Message{Type: t, Event: ev}, true, nil
	case MessageTypeAudioChunk:
		c, err := decodeAudioChunk(payload)
		if err != nil {
			return Message{}, false, err
		}
		return Message{Type: t, AudioChunk: c}, true, nil
	case MessageTypePong:
		p, err := decodePong(payload)
		if err != nil {
			return Message{}, false, err
		}
		return Message{Type: t, Pong: p}, true, nil
	case MessageTypeTrace:
		tr, err := decodeTraceRecord(payload)
		if err != nil {
			return Message{}, false, err
		}
		return Message{Type: t, Trace: tr}, true, nil
	default:
		return Message{}, false, nil
	}
}

// readMessage reads frames from r until one decodes to a known
// MessageType, skipping any unknown type by the length its header
// declares.
func readMessage(r io.Reader) (Message, error) {
	for {
		t, payload, err := readFrame(r)
		if err != nil {
			return Message{}, err
		}
		msg, ok, err := decodeMessage(t, payload)
		if err != nil {
			return Message{}, err
		}
		if ok {
			return msg, nil
		}
	}
}

func encodeEvent(ev Event) []byte {
	buf := make([]byte, eventSize)
	buf[0] = ev.KeyIndex
	buf[1] = byte(ev.Type)
	binary.LittleEndian.PutUint64(buf[2:10], ev.Timestamp)
	return buf
}

func decodeEvent(payload []byte) (Event, error) {
	if len(payload) != eventSize {
		return Event{}, fmt.Errorf("transport: event payload is %d bytes, want %d", len(payload), eventSize)
	}
	return Event{
		KeyIndex:  payload[0],
		Type:      EventType(payload[1]),
		Timestamp: binary.LittleEndian.Uint64(payload[2:10]),
	}, nil
}

// encodeAudioChunk builds an audio chunk's frame payload: stream ID, PCM
// samples, then the final-chunk flag. The chunk's length is no longer
// self-described; the frame header carries it.
func encodeAudioChunk(c AudioChunk) []byte {
	buf := make([]byte, 0, 2+len(c.PCM))
	buf = append(buf, c.StreamID)
	buf = append(buf, c.PCM...)
	final := byte(0)
	if c.Final {
		final = 1
	}
	return append(buf, final)
}

func decodeAudioChunk(payload []byte) (AudioChunk, error) {
	if len(payload) < 2 {
		return AudioChunk{}, fmt.Errorf("transport: audio chunk payload is %d bytes, want at least 2", len(payload))
	}
	return AudioChunk{
		StreamID: payload[0],
		PCM:      payload[1 : len(payload)-1],
		Final:    payload[len(payload)-1] != 0,
	}, nil
}

// encodePong builds a Pong frame's payload: the nonce, unchanged from the
// ping that carried it.
func encodePong(nonce byte) []byte {
	return []byte{nonce}
}

func decodePong(payload []byte) (Pong, error) {
	if len(payload) != pongSize {
		return Pong{}, fmt.Errorf("transport: pong payload is %d bytes, want %d", len(payload), pongSize)
	}
	return Pong{Nonce: payload[0]}, nil
}

// encodeTraceRecord builds a Trace record frame's payload: code, key,
// payload, then the device's monotonic timestamp, matching
// firmware/trace.py's Tracer.
func encodeTraceRecord(tr TraceRecord) []byte {
	buf := make([]byte, traceRecordSize)
	buf[0] = byte(tr.Code)
	buf[1] = tr.Key
	binary.LittleEndian.PutUint16(buf[2:4], tr.Payload)
	binary.LittleEndian.PutUint64(buf[4:12], tr.Timestamp)
	return buf
}

func decodeTraceRecord(payload []byte) (TraceRecord, error) {
	if len(payload) != traceRecordSize {
		return TraceRecord{}, fmt.Errorf("transport: trace record payload is %d bytes, want %d", len(payload), traceRecordSize)
	}
	return TraceRecord{
		Code:      TraceCode(payload[0]),
		Key:       payload[1],
		Payload:   binary.LittleEndian.Uint16(payload[2:4]),
		Timestamp: binary.LittleEndian.Uint64(payload[4:12]),
	}, nil
}

File: firmware/wire.py
"""Encodes and decodes the messages in docs/wire-protocol.md.

This is the firmware half of the contract that `driver/transport/wire.go`
implements on the host side. Both files build against the same document,
so a change to one message's layout changes both, and both refuse to
guess at field offsets when the version byte does not match.

See tasks/ongoing/0022-firmware-main-loop.md for the design decision.
"""

PROTOCOL_VERSION = 1

KEY_STATE_SIZE = 6  # docs/wire-protocol.md's Key state message
EVENT_SIZE = 10  # docs/wire-protocol.md's Press/release event
FRAME_HEADER_SIZE = 3  # docs/wire-protocol.md's Framing: type + length

PRESS = 0
RELEASE = 1

MESSAGE_TYPE_EVENT = 1  # docs/wire-protocol.md's Framing type registry
MESSAGE_TYPE_PONG = 3  # docs/wire-protocol.md's Framing type registry
MESSAGE_TYPE_TRACE = 4  # docs/wire-protocol.md's Framing type registry
MESSAGE_TYPE_SET_CUSTOM_GLYPH = 5  # docs/wire-protocol.md's Framing type registry

PING_KEY_INDEX = 255  # docs/wire-protocol.md's Ping section

CUSTOM_GLYPH_WIDTH = 128
CUSTOM_GLYPH_HEIGHT = 128
# docs/wire-protocol.md's "Set custom glyph": 128x128 pixels, RGB565, 2
# bytes each.
CUSTOM_GLYPH_PIXELS_SIZE = CUSTOM_GLYPH_WIDTH * CUSTOM_GLYPH_HEIGHT * 2
CUSTOM_GLYPH_PAYLOAD_SIZE = 1 + CUSTOM_GLYPH_PIXELS_SIZE  # key index + pixels

# docs/wire-protocol.md's "Emoji IDs": a key showing its last custom
# image, not a built-in glyph table entry.
CUSTOM_GLYPH_SENTINEL_EMOJI_ID = 0xFE


class UnsupportedVersionError(ValueError):
    """A key state message carried a version byte this build cannot read.

    Mirrors `transport.ErrUnsupportedVersion` on the driver side. It
    subclasses `ValueError` so one `except ValueError` covers both this
    and a short or malformed message.
    """


class KeyState:
    """One decoded key state message from the host.

    `emoji_id` is the wire's 1-byte index into the firmware's glyph
    table, not a glyph itself — task 0023 supplies the lookup that turns
    it into a bitmap.
    """

    def __init__(self, key_index, version, color, emoji_id, blink):
        self.key_index = key_index
        self.version = version
        self.color = color
        self.emoji_id = emoji_id
        self.blink = blink


def decode_key_state(data):
    """Decode a 6-byte key state message sent by the host over HID.

    Raises `UnsupportedVersionError` when the version byte is not
    `PROTOCOL_VERSION`, and `ValueError` when the message is short. The
    version check comes before any field read after it, because a format
    change moves the offsets of `color`, `emoji_id`, and `blink`.
    """
    if len(data) != KEY_STATE_SIZE:
        raise ValueError(
            "key state message is {} bytes, want {}".format(
                len(data), KEY_STATE_SIZE
            )
        )

    version = data[1]
    if version != PROTOCOL_VERSION:
        raise UnsupportedVersionError(
            "got version {}, want {}".format(version, PROTOCOL_VERSION)
        )

    return KeyState(
        key_index=data[0],
        version=version,
        color=data[2] | (data[3] << 8),
        emoji_id=data[4],
        blink=data[5] != 0,
    )


def encode_event(key_index, event_type, timestamp_us):
    """Build the 10-byte press or release event the host reads from CDC.

    `event_type` is `PRESS` or `RELEASE`. `timestamp_us` is monotonic
    microseconds, written little-endian to match the RP2350's byte order.
    """
    if event_type not in (PRESS, RELEASE):
        raise ValueError("event type {} is not PRESS or RELEASE".format(event_type))

    buffer = bytearray(EVENT_SIZE)
    buffer[0] = key_index
    buffer[1] = event_type
    buffer[2:EVENT_SIZE] = timestamp_us.to_bytes(8, "little")
    return bytes(buffer)


def write_frame(writer, message_type, payload):
    """Write one device→host frame: a type byte, a little-endian uint16
    payload length, then the payload itself, to `writer`.

    Mirrors `driver/transport/wire.go`'s `writeFrame`, so `Device.ReadMessage`
    on the host side can decode what this writes. See "Framing" in
    docs/wire-protocol.md.
    """
    header = bytes((message_type, len(payload) & 0xFF, (len(payload) >> 8) & 0xFF))
    writer.write(header)
    writer.write(payload)


def encode_pong(nonce):
    """Build a Pong reply's payload: `nonce`, unchanged.

    Pass this to `write_frame` with `MESSAGE_TYPE_PONG` to write the
    framed reply a caller reads back with `driver/transport`'s Pong
    decoder.
    """
    return bytes((nonce,))


class CustomGlyph:
    """One decoded Set custom glyph message: a key index and its raw
    128x128 RGB565 pixel buffer. See "Set custom glyph" in
    docs/wire-protocol.md.
    """

    def __init__(self, key_index, pixels):
        self.key_index = key_index
        self.pixels = pixels


def decode_custom_glyph(payload):
    """Decode a Set custom glyph message's frame payload.

    Raises `ValueError` when `payload` is not exactly
    `CUSTOM_GLYPH_PAYLOAD_SIZE` bytes.
    """
    if len(payload) != CUSTOM_GLYPH_PAYLOAD_SIZE:
        raise ValueError(
            "custom glyph payload is {} bytes, want {}".format(
                len(payload), CUSTOM_GLYPH_PAYLOAD_SIZE
            )
        )
    return CustomGlyph(key_index=payload[0], pixels=bytes(payload[1:]))


class CustomGlyphReader:
    """Incrementally reads one framed CDC message from the host.

    A Set custom glyph payload is up to 32,769 bytes — too large to read
    in one `MacroPad.step` without stalling the switch scan and event
    loop — so this buffers whatever `reader.in_waiting` reports on each
    `feed` call, across as many calls as it takes, instead of blocking on
    a full frame arriving at once.
    """

    def __init__(self):
        self._buffer = bytearray()

    def feed(self, reader):
        """Read whatever is available from `reader`.

        Returns a decoded `CustomGlyph` once a full Set custom glyph
        frame has arrived, `None` otherwise. A frame of a type this
        reader does not know is dropped once fully buffered, by its
        declared length, matching `driver/transport/wire.go`'s
        `readFrame`/`decodeMessage` split.
        """
        available = reader.in_waiting
        if available:
            self._buffer.extend(reader.read(available))

        if len(self._buffer) < FRAME_HEADER_SIZE:
            return None

        message_type = self._buffer[0]
        length = self._buffer[1] | (self._buffer[2] << 8)
        total = FRAME_HEADER_SIZE + length
        if len(self._buffer) < total:
            return None

        payload = bytes(self._buffer[FRAME_HEADER_SIZE:total])
        del self._buffer[:total]

        if message_type != MESSAGE_TYPE_SET_CUSTOM_GLYPH:
            return None
        return decode_custom_glyph(payload)
